File: bench/testdata/samples/A07/PSB-A07-JWT-001/python/secure.py
# PSB-A07-JWT-001: JWT authentication middleware
# CWE: CWE-345, CWE-347
# Expected: (none - secure)
from flask import Flask, request, jsonify, g
import jwt
import os
import datetime
from functools import wraps

app = Flask(__name__)
SECRET_KEY = os.environ["JWT_SECRET"]

def auth_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        header = request.headers.get("Authorization", "")
        if not header.startswith("Bearer "):
            return jsonify({"error": "missing authorization header"}), 401

        token = header[7:]
        try:
            decoded = jwt.decode(
                token, SECRET_KEY, algorithms=["HS256"],
                options={"require": ["exp", "sub"]},
            )
            g.user = decoded
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "token expired"}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "invalid token"}), 401
        return f(*args, **kwargs)
    return decorated

@app.route("/api/login", methods=["POST"])
def login():
    data = request.get_json()
    token = jwt.encode(
        {"sub": data["username"], "role": "user",
         "exp": datetime.datetime.utcnow() + datetime.timedelta(hours=1)},
        SECRET_KEY, algorithm="HS256",
    )
    return jsonify({"token": token})

@app.route("/api/profile")
@auth_required
def profile():
    return jsonify({"user": g.user})

if __name__ == "__main__":
    app.run()

File: bench/testdata/samples/A07/PSB-A07-JWT-001/javascript/secure.js
// PSB-A07-JWT-001: JWT authentication middleware
// CWE: CWE-345, CWE-347
// Expected: (none - secure)
const express = require("express");
const jwt = require("jsonwebtoken");
const app = express();
app.use(express.json());

const SECRET = process.env.JWT_SECRET;

function authMiddleware(req, res, next) {
  const header = req.headers.authorization;
  if (!header || !header.startsWith("Bearer ")) {
    return res.status(401).json({ error: "missing authorization header" });
  }

  const token = header.slice(7);
  try {
    const decoded = jwt.verify(token, SECRET, {
      algorithms: ["HS256"],
      complete: true,
    });
    req.user = decoded.payload;
    next();
  } catch (err) {
    res.status(401).json({ error: "invalid token" });
  }
}

app.post("/api/login", (req, res) => {
  const { username, password } = req.body;
  const token = jwt.sign({ sub: username, role: "user" }, SECRET, {
    algorithm: "HS256",
    expiresIn: "1h",
  });
  res.json({ token });
});

app.get("/api/profile", authMiddleware, (req, res) => {
  res.json({ user: req.user });
});

module.exports = app;
